### PR TITLE
feat(planner): Support nested joins

### PR DESCRIPTION
**Pull request by jon-chuang**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?
support nested joins

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### Release note
Nested joins in select statements.

## Refer to a related PR or issue link (optional)
Replace: https://github.com/singularity-data/risingwave/pull/3852
Closes: https://github.com/singularity-data/risingwave/issues/3816
Closes: https://github.com/singularity-data/risingwave/issues/4418

### CODE REVIEW

File: src/frontend/src/binder/bind_context.rs [hunk 1]
```diff
@@ -235,7 +235,7 @@ impl BindContext {
                 if group.non_nullable_column.is_none() {
                     group.non_nullable_column = non_nullable_column;
                 }
-                self.column_group_context.mapping.insert(right, group_id);
```
Comment (jon-chuang) [inline src/frontend/src/binder/bind_context.rs:238]: This was a bug, its now fixed

File: src/frontend/src/binder/bind_context.rs [hunk 2]
```diff
@@ -235,7 +235,7 @@ impl BindContext {
...
+                self.column_group_context.mapping.insert(left, group_id);
```
Comment (jon-chuang) [inline src/frontend/src/binder/bind_context.rs:238]: This was a bug, its now fixed

**Comment by jon-chuang** [comment]
Hello, this PR has been ready for a while, would someone kindly take a look? :)

File: src/frontend/test_runner/tests/testdata/join.yaml [hunk 3]
```diff
@@ -507,6 +536,27 @@
                 StreamTableScan { table: a, columns: [a.x, a._row_id], pk: [a._row_id], distribution: HashShard(a._row_id) }
               StreamExchange { dist: HashShard(b.x) }
                 StreamTableScan { table: b, columns: [b.x, b._row_id], pk: [b._row_id], distribution: HashShard(b._row_id) }
+- sql: |
+    /* Test that we can join subqueries without aliases */
+    select * from (select 1 as x) join (select 1 as x) using(x);
```
Comment (chenzl25) [inline src/frontend/test_runner/tests/testdata/join.yaml:541]: Subqueries without aliases seems not a correct syntax？

Comment (chenzl25) [reply src/frontend/test_runner/tests/testdata/join.yaml:541]: At least, both MySQL and PostgreSQL don't consider this sql as a valid sql.
They both think every derived table must have its own alias.

Comment (jon-chuang) [reply src/frontend/test_runner/tests/testdata/join.yaml:541]: Yes, it's not supported in those. 

Maybe I'll post in the slack channel to ask what people think we should follow. But for instance values also has no alias, and at least for now I don't see a practical reason why we should not support subquery in from statement without alias.

Comment (jon-chuang) [reply src/frontend/test_runner/tests/testdata/join.yaml:541]: But anyway, it's not a problem not to support it either. I can remove this test and in another PR we can enforce that FROM subquery always has an alias in parser

**Review by chenzl25 — APPROVED** [review]
LGTM

**Comment by fuyufjh** [comment]
hi, any updates?

**Comment by jon-chuang** [comment]
yes, there was an error from merging main, now fixed.

**Pull request merged**